src/main/java/io/cryostat/platform/ServiceRef.java

```diff
@@ -218,4 +220,74 @@ public enum AnnotationKey {
         REALM,
         ;
     }
+
+    public static Compare compare(Collection<ServiceRef> src) {
+        return new Compare(src);
+    }
+
+    public static class Compare {
+        private Collection<ServiceRef> previous, current;
+
+        public Compare(Collection<ServiceRef> previous) {
+            this.previous = new HashSet<>(previous);
+        }
+
+        public Compare to(Collection<ServiceRef> current) {
+            this.current = new HashSet<>(current);
+            return this;
+        }
+
+        public Collection<ServiceRef> added() {
+            return removeAllUpdatedRefs(addedOrUpdatedRefs(), updated());
+        }
+
+        public Collection<ServiceRef> removed() {
+            return removeAllUpdatedRefs(removedOrUpdatedRefs(), updated());
+        }
+
+        public Collection<ServiceRef> updated() {
+            Collection<ServiceRef> added = addedOrUpdatedRefs();
+            Collection<ServiceRef> removed = removedOrUpdatedRefs();
+            Collection<ServiceRef> updated = new HashSet<>();
+
+            // Manual Collection intersection since ServiceRef also compares jvmId
+            for (ServiceRef addedRef : added) {
+                for (ServiceRef removedRef : removed) {
+                    if (Objects.equals(addedRef.getServiceUri(), removedRef.getServiceUri())) {
+                        updated.add(addedRef);
+                    }
+                }
+            }
+
+            return updated;
+        }
+
+        private Collection<ServiceRef> addedOrUpdatedRefs() {
+            Collection<ServiceRef> added = new HashSet<>(current);
+            added.removeAll(previous);
+            return added;
+        }
+
+        private Collection<ServiceRef> removedOrUpdatedRefs() {
+            Collection<ServiceRef> removed = new HashSet<>(previous);
+            removed.removeAll(current);
+            return removed;
+        }
+
+        private Collection<ServiceRef> removeAllUpdatedRefs(
+                Collection<ServiceRef> src, Collection<ServiceRef> updated) {
+            Collection<ServiceRef> tnSet = new HashSet<>(src);
+
+            // Manual removal since ServiceRef also compares jvmId
+            for (ServiceRef srcRef : src) {
+                for (ServiceRef updatedRef : updated) {
+                    if (Objects.equals(srcRef.getServiceUri(), updatedRef.getServiceUri())) {
```

Comment:
Maybe this for-loop can be extracted to a private helper method since it looks like the loop implementation is the same here and above.